lib/error_report_notify.c

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * Userspace notification interface for debugging tools.
+ *
+ * Provide two sysfs files:
+ *  - /sys/kernel/error_report/last_report
+ *  - /sys/kernel/error_report/report_count
+ * that contain the last debugging tool report (taken from dmesg, delimited by
+ * the error_report_start/error_report_end tracing events) and the total report
+ * count.
+ *
+ * Userspace apps can call poll(POLLPRI) on those files to get notified about
+ * the new reports without having to watch dmesg in a loop.
+ *
+ * Because reporting code can be called from low-level routines (e.g. locking
+ * primitives or allocator guts), report recording is implemented using a
+ * seqlock-alike lock-free algorithm.
+ *
+ */
+
+#include <linux/fs.h>
+#include <linux/string.h>
+#include <linux/tracepoint.h>
+#include <linux/vmalloc.h>
+#include <linux/workqueue.h>
+#include <trace/events/error_report.h>
+#include <trace/events/printk.h>
+
+static struct kobject *error_report_kobj;
+
+#define BUF_SIZE PAGE_SIZE
```

Comment:
I think we can't have more than PAGE_SIZE because of vfs file size limit, right? If so, a brief comment here would be good.

Reply:
Yes, you are right. Added the comment.